### ISSUE TITLE
Correct handling of interior boundaries (related to "land masks")

### ISSUE THREAD

**Issue by rabernat** [xgcm/xgcm#324]
I wanted to raise an issue that has come up via @stb2145's work with SODA. 

The basic issue is that `interp` and `diff` are not aware of "interior boundaries", i.e. grid cells that are next to land. In the presence of such boundaries, the methods need to modify their weights to account for the "missing" data, applying some sort of boundary condition (e.g. Dirichlet or Neumann). Without explicit knowledge of land masks, the current implementation either grows the mask on the input data (in the case of NaN-filled data) or else gives a wrong answer near boundaries (in the case of 0- or constant-filled data). That's why I label this a bug.

I put together an example of a workaround outside of xgcm here:
https://nbviewer.jupyter.org/gist/rabernat/ade98c74909271a0d4dd8f7d7c435640

Basically this workaround uses the sort of kernel-based approach we are using in gcm-filters, also prototyped for xgcm in https://github.com/ocean-transport/coiled_collaboration/pull/12.

Shanice: to apply this method on dask-backed data, you would want to use `apply_ufunc` rather than the direct call to the `interp_right_with_land` function shown in the notebook.

Going forward, we need to decide if we want to refactor xgcm to allow for these sorts of kernel functions.

(Discussed tangentially in #293)

**Comment by rabernat** [comment]
I have updated the gist with an example of using this via apply_ufunc:

https://nbviewer.jupyter.org/urls/gist.githubusercontent.com/rabernat/ade98c74909271a0d4dd8f7d7c435640/raw/726d9e3a80b08875d6aa959252388a3794a78fc9/land-aware-interp.ipynb

**Comment by jbusecke** [comment]
This is marked as 'bug', why is the stale bot doing stuff here?

**Comment by jbusecke** [comment]
I think it is worth starting a chat about how to implement this on a high level, since we will soon have an opportunity to implement this similarly to what @rabernat has done above with #344.

On a very general level this involves 'filling' some grid points with values. This might have to involve adding the concept of a 'mask' similar to 'metrics' to xgcm (#293).

I am  proposing the API should look something like this: 
```python
grid.interp(..., internal_boundary='something')
```
Do folks here have thoughts about this signature?

I think this will work nicely from a performance standpoint, since we could wrap the  filling within the ufunc step in the former, but not in the latter. We can discuss later how to exactly do this (e.g. writing a ton of new ufuncs or adding a kwargs and a call to a separate function in the gridops.py module).

In terms of terminology, I am not a big fan of `internal_boundary`, but for the time being we need to separate physical boundaries (e.g. land) and domain boundaries (what to do with the edge of the array). Does anyone have suggestions for a more intuitive name?

**Comment by rabernat** [comment]
> Do folks here have thoughts about this signature?

I don't like `internal_boundary`. The grid_ufunc approach will bring some changes to how we think about boundaries. The Grid will now handle the "global array boundary" (i.e. what to do at the edges of the array) automatically, based on the grid topology. (For non-periodic Grids, one option may be to create a virtual land mask for points outside of the domain).

The only true "boundary conditions" from a PDE point of view will be the "internal" boundaries, i.e. cells that border some sort of land mask. Thus it should be simple enough to have just one "boundary" argument.

We would need different grid_ufuncs in the presence of such boundaries. With no land, we just have


```python
@as_grid_ufunc(signature="(X:center)->(X:left)", boundary_width={"X": (1, 0)})
def interp_center_to_left(a):
    return 0.5 * (a[..., 1:] + a[..., :-1])
```

With land, we need another argument:
```python
@as_grid_ufunc(signature="(X:center),(X:center)->(X:left)", boundary_width={"X": (1, 0)})
def interp_center_to_left_w_mask(a, wet_mask):
    # wet_fac = 0: water on left AND right
    # wet_fac = 1: water on left OR right
    # wet_fac = 2: no land nearby
    wet_fac = wet_mask[..., 1:] + wet_mask[..., :-1]
    return (a[..., 1:] + a[..., :-1]) / wet_fac
```

In this sense `wet_mask` is a property of the Grid, like the metrics. It would be great to have a way for the Grid to automatically fill in such arguments when the grid_ufunc is invoked.

**Comment by dcherian** [comment]
The explosion of ufuncs feels wrong to me. 

In this case though, can a grid not have `wet_mask=1` (appropriate DataArrays actually) by default? In which case xgcm expects all ufuncs to handle it.  This makes sense given this point:
> In this sense wet_mask is a property of the Grid, like the metrics.

**Comment by rabernat** [comment]
> The explosion of ufuncs feels wrong to me.

Can you say more about why?

There are performance tradeoffs. The first ufunc will be faster and allocate less memory. If we don't care about land masks, it's unambiguously better. I have no problem writing lots of different ufuncs to cover different special cases. The point of the protocol is to eliminate the boilerplate code so we can focus on the core computation.

The main challenge I see will be figuring out which ufunc to dispatch given our current API.

**Comment by dcherian** [comment]
Because it instantly doubles the number of ufuncs anyone else has to write if they're doing something custom about land edges.

> The first ufunc will be faster and allocate less memory.

Sure, but to check whether `wet_mask` is all True takes 8ms in a 170ms calculation at the upper end of chunksizes we tend to recommend. Is it really worth increasing the amount of code by ~ 30%?

``` python
# 400MB data chunk
data = np.ones((100, 1000, 1000), dtype=np.float32)
wet_mask = np.ones((100, 1000, 1000), dtype=bool)


def interp(data, wet_mask):
    if not wet_mask.all():
        int_mask = wet_mask.astype(np.int8)
        data = int_mask * data
        wfac = int_mask[..., :-1] + int_mask[..., 1:]
        return (data[..., :-1] + data[..., 1:]) / wfac
    else:
        return (data[..., :-1] + data[..., 1:]) / 2


%timeit wet_mask.all()
%timeit interp(data, wet_mask)
wet_mask[-1] = False
%timeit interp(data, wet_mask)
```
```
8.66 ms ± 385 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
174 ms ± 10.9 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)

/tmp/ipykernel_1264027/2687882246.py:10: RuntimeWarning: invalid value encountered in true_divide
  return (data[..., :-1] + data[..., 1:]) / wfac

349 ms ± 20.5 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

**Comment by TomNicholas** [comment]
From a code simplicity point of view it would be nice to find a way to do this generally (and optionally) within `apply_as_grid_ufunc`.

**Comment by jbusecke** [comment]
First of all, thanks for all the comments. These are super helpful.

> I don't like internal_boundary. The grid_ufunc approach will bring some changes to how we think about boundaries. The Grid will now handle the "global array boundary" (i.e. what to do at the edges of the array) automatically, based on the grid topology. (For non-periodic Grids, one option may be to create a virtual land mask for points outside of the domain).

Neither do I. If we introduce breaking changes we could switch to something like `pad` for the array boundaries (in line with the functionality we will be putting in soon) and use `boundary` for physical boundaries only?

From a performance point, could we not shift this logic to the grid creation?
```python
if not wet_mask.all():
   int_mask = wet_mask.astype(np.int8)
```
in most cases this will be some 2d field and the overhead to grid creation should be minimal, but saving us the computation for each of the computation steps.
The grid would then carry the information whether there is an 'applyable' mask and if not we could just default to `None`, and the ufuncs could skip the filling step altoghether? 
In a naive setup:
```python

def diff_forward(a):
    return a[..., 1:] - a[..., :-1]

def apply_mask(a, mask, physical_boundary):
   # fill values with appropriate method
   return filled_a


@as_grid_ufunc(signature="(X:center)->(X:left)", mask=None,physical_boundary='something', boundary_width={"X": (1, 0)})
def diff_center_to_left(a):
    if mask:
        a = apply_mask(a, mask, physical_boundary)
    return diff_forward(a)


@as_grid_ufunc(signature="(X:left)->(X:center)", mask=None, physical_boundary='something', boundary_width={"X": (0, 1)})
def diff_left_to_center(a):
    if mask:
        a = apply_mask(a, mask, physical_boundary)
    return diff_forward(a)

```
If I am not missing something here, this would not be significantly slower when `mask=None`? I think @TomNicholas s design specifically allowed for more kwargs added this way?

I think overall this discussion is very valuable, because I anticipate that we might add some more inputs that are similar to metrics (e.g. partial cells) and then the increase in ufuncs would become really unwieldy.

**Comment by rabernat** [comment]
Really enjoying the discussion. Everyone has such good points. 

I think we are focused on different aspects that are not always mutually compatible. I am thinking about performance above all, since that is the reason we did the whole massive refactor in the first place--I want the grid_ufuncs to as fast as possible at the numpy level in every different scenario. Deepak seems to be thinking about the xgcm user experience. Tom is thinking about the code architecture part. Julius is trying to reconcile all of the above.

To clarify my suggestion, I am not proposing that xgcm _users_ would have to know the difference between `interp_center_to_left` and `interp_center_to_left_w_mask`. I am envisioning an extra layer of "dispatch" between `Grid.interp` and these lower level ufuncs. I have no problem with _us_ implementing a bunch of specialized grid_ufuncs for different circumstances, if there are performance advantages. But users should still just be able to call `Grid.interp` and have it do the "right" thing given their grid.

Specifically, I am assuming the initialized `Grid` object will already know whether it has a `wet_mask` or not. Then `.apply_grid_ufunc` will _somehow_* know whether to call  `interp_center_to_left` or `interp_center_to_left_w_mask`. So we skip even the `if not wet_mask.all()` call. Deepak's timing in https://github.com/xgcm/xgcm/issues/324#issuecomment-988246258 does also show that the difference between the two grid_ufuncs is nearly 2x, so picking the right one is important. Could we write them in numba and get 10x?

This is essentially a multiple-dispatch style problem, and as such it is a bit awkward in python. But I think we can make it work at least internally within xgcm for our specific use cases. The key question is: how does `apply_as_grid_ufunc` figure out which method to use under which scenario?

Maybe we need some sort of higher order concept of a grid_ufunc which has rules about dispatching to other private, more specialized methods? That sounds complicated though.

Deepak's comment also reveals the value of benchmarking, which we should do regardless of how we implement.

**Comment by jbusecke** [comment]
Thanks for the detailed comment, Ryan. 

I ultimately think that we can make this discussion more quantitive by following this suggestion first?
> Deepak's comment also reveals the value of benchmarking, which we should do regardless of how we implement.

Should we try to start benchmarking the code somehow before merging @TomNicholas refactor?

For this effort to be transparent I think we need to be a bit more precise with the meaning of "performance" here. 

To me there are two aspects to the overall performance: The amount of dask chunks created (the main change of @TomNicholas refactor AFAIKT) and the amount of time each chunks needs for execution (@dcherian benchmark above). These can have non-trivial effects on high level method performance (e.g. when the chunking is suboptimal I can see a case where improving the ufunc speed has marginal influence on overall performance due to the dask task overhead?)
I believe we should test/benchmark each of these individually + test a 'typical' high level computation on a few sizes of datasets. 

Having these datapoints for each PR would give us some less speculative ground to make a decision on. This is however a substantial amount of work? 

Some more comments on the above:

> will somehow* know whether to call interp_center_to_left or interp_center_to_left_w_mask. 

As far as I understood, Toms dispatch is already pretty flexible and could probably handle this already (@TomNicholas correct me if I am wrong here). 

>  I have no problem with us implementing a bunch of specialized grid_ufuncs for different circumstances, if there are performance advantages.

I would like to comment on this and would lean towards @dcherian here, because this could get out of hand rather quickly even for us maintainers.
From our discussion I always had the impression that if the user wants to `squeeze the last bit of performance` out of a calculation they would write their own specialized ufunc? We want the 'stock' ufuncs to be fast, but if we can save hundreds of lines of code and sacrifice only a few percent performance, I would personally go that route.
Connected to this: As I said above, I think we will not at all need the `.all()` call in the ufuncs. So if all that we add is `if mask:` will that actually affect the performance?

**Comment by TomNicholas** [comment]
> Should we try to start benchmarking the code somehow before merging @TomNicholas refactor?

The only "benchmarking" I have done so far was checking that the new dask algorithm creates the expected graph, same as what Ryan did in his notebook. I have a ToDo somewhere to at least write a test which checks if a change accidentally increases the number of required dask tasks. I have not really been thinking about the speed of the ufuncs themselves, partly because I was trying to do it in such a way that the ufunc could be anything, and also because my understanding was that dask graph performance at scale was the primary motivation (so that our group can handle LLC4320 data specifically).

> As far as I understood, Tom's dispatch is already pretty flexible and could probably handle this already (@TomNicholas correct me if I am wrong here).

FYI [This is where the ufunc dispatch code lives at the moment](https://github.com/xgcm/xgcm/blob/f7599172bd7bb262ea14adecfe018220441cd23a/xgcm/grid.py#L2231) - right now it just looks for any grid ufuncs defined in the `gridops.py` module which match in name and signature, but you can see the pseudocode intended to match any kwarg.

The way I was imagining the dispatch working was to select via the kwargs given to the `as_grid_ufunc` decorator, but still allow arbitrary kwargs to be passed to the function. That would mean there was a difference between this

```python
@as_grid_ufunc(signature="(X:center)->(X:left)", boundary_width={"X": (1, 0)})
def interp_center_to_left(arr, wet_mask=DEFAULT):
    ...
```

which means the one `interp_center_to_left` function needs to handle both wet and dry cases so to speak, and this

```python
@as_grid_ufunc(signature="(X:center)->(X:left)", boundary_width={"X": (1, 0)}, wet_mask=DEFAULT)
def interp_center_to_left(arr):
    ...
```

which could allow for two different `interp_center_to_left` ufuncs, which are chosen between based on the presence of the `wet_mask` kwarg in the user's call.

A third option is to have the latter syntax, but instead of dispatching on `wet_mask` it gets handled generically in some way inside `apply_as_grid_ufunc`.

> This is essentially a multiple-dispatch style problem, and as such it is a bit awkward in python.

There is more than one library for multiple dispatch in python, but I doubt that's the right solution here...

**Comment by TomNicholas** [comment]
Also I think we should concentrate on making this work in the simplest way first, and then if that turns out not to be fast enough we make more granular ufuncs later. That way we get the benefit of improved dask graphs, but don't increase code complexity unless we need to.

**Comment by rabernat** [comment]
I'm fine with that approach.
